src/SA/listmods/entry.c

```diff
@@ -0,0 +1,130 @@
+#include <windows.h>
+#include <stdio.h>
+#include "psapi.h"
+#include "bofdefs.h"
+#include "base.c"
+
+DECLSPEC_IMPORT LPVOID WINAPI KERNEL32$OpenProcess(DWORD dwDesiredAccess, BOOL bInheritHandle, DWORD dwProcessId);
+DECLSPEC_IMPORT DWORD KERNEL32$GetCurrentProcessId();
+DECLSPEC_IMPORT HGLOBAL KERNEL32$GlobalAlloc(UINT uFlags, SIZE_T dwBytes);
+DECLSPEC_IMPORT HGLOBAL KERNEL32$GlobalFree(HGLOBAL hMem);
+DECLSPEC_IMPORT WINBOOL WINAPI PSAPI$EnumProcessModulesEx(HANDLE hProcess, HMODULE *lphModule, DWORD cb, LPDWORD lpcbNeeded, DWORD dwFilterFlag);
+DECLSPEC_IMPORT DWORD PSAPI$GetModuleFileNameExA(HANDLE hProcess, HMODULE hModule, LPSTR lpFilename, DWORD nSize);
+
+typedef DWORD GetFileVersionInfoSizeA_t(
+  LPCSTR  lptstrFilename,
+  LPDWORD lpdwHandle
+);
+
+typedef WINBOOL GetFileVersionInfoA_t(
+  LPCSTR lptstrFilename,
+  DWORD  dwHandle,
+  DWORD  dwLen,
+  LPVOID lpData
+);
+
+typedef WINBOOL VerQueryValueA_t(
+  LPCVOID pBlock,
+  LPCSTR  lpSubBlock,
+  LPVOID  *lplpBuffer,
+  PUINT   puLen
+);
+
+// Inspired from: https://www.codeguru.com/cpp/w-p/win32/versioning/article.php/c4539/Versioning-in-Windows.htm
+int PrintSingleModule(char* szFile){
+    DWORD dwLen, dwUseless;
+    LPTSTR lpVI;
+    char* companyName;
+
+    GetFileVersionInfoSizeA_t* GetFileVersionInfoSizeA = (GetFileVersionInfoSizeA_t*) GetProcAddress(LoadLibraryA("Api-ms-win-core-version-l1-1-0.dll"), "GetFileVersionInfoSizeA");
```

Comment:
For these please use the Beacon Object file syntax for calling these functions.

If you need help getting the definitions I outline my workflow in [https://www.trustedsec.com/blog/a-developers-introduction-to-beacon-object-files/](https://www.trustedsec.com/blog/a-developers-introduction-to-beacon-object-files/)

this function for example would have this definition added to bofdefs.h
DWORD WINAPI VERSION$GetFileVersionInfoSizeA(LPCSTR lptstrFilename,LPDWORD lpdwHandle);

and would be called using VERSION$GetFileVersionInfoSizeA(szFile, &dwUseless);

Reply:
Done. Thanks for the hint. Does that mean that the beacon resolve based on the library name and not the DLL name?

Reply:
It is still resolving by dll, what you ran into is a weird "virtual dll" sublayer that windows uses. you can read more about that here if interested [https://www.geoffchappell.com/studies/windows/win32/apisetschema/index.htm](https://www.geoffchappell.com/studies/windows/win32/apisetschema/index.htm)

In this case the lib name happens to associate with the backing dll.  If you check your windows system for version.dll under system32, you'll find that exists, where the api-...-.dll doesn't.

And just for a self check using dumpbin or objdump on version.dll will show these results for the exports from that dll.
```
The Export Tables (interpreted .rdata section contents)

Export Flags 			0
Time/Date stamp 		b11b88e5
Major/Minor 			0/0
Name 				0000000000004cb2 VERSION.dll
Ordinal Base 			1
Number in:
	Export Address Table 		00000011
	[Name Pointer/Ordinal] Table	00000011
Table Addresses
	Export Address Table 		0000000000004c08
	Name Pointer Table 		0000000000004c4c
	Ordinal Table 			0000000000004c90

Export Address Table -- Ordinal Base 1
	[   0] +base[   1] 1080 Export RVA
	[   1] +base[   2] 2190 Export RVA
	[   2] +base[   3] 1df0 Export RVA
	[   3] +base[   4] 1040 Export RVA
	[   4] +base[   5] 1010 Export RVA
	[   5] +base[   6] 1e00 Export RVA
	[   6] +base[   7] 1050 Export RVA
	[   7] +base[   8] 1060 Export RVA
	[   8] +base[   9] 1070 Export RVA
	[   9] +base[  10] 1e10 Export RVA
	[  10] +base[  11] 2360 Export RVA
	[  11] +base[  12] 1e20 Export RVA
	[  12] +base[  13] 2f80 Export RVA
	[  13] +base[  14] 4ddc Forwarder RVA -- KERNEL32.VerLanguageNameA
	[  14] +base[  15] 4e07 Forwarder RVA -- KERNEL32.VerLanguageNameW
	[  15] +base[  16] 1020 Export RVA
	[  16] +base[  17] 1030 Export RVA

[Ordinal/Name Pointer] Table
	[   0] GetFileVersionInfoA
	[   1] GetFileVersionInfoByHandle
	[   2] GetFileVersionInfoExA
	[   3] GetFileVersionInfoExW
	[   4] GetFileVersionInfoSizeA
	[   5] GetFileVersionInfoSizeExA
	[   6] GetFileVersionInfoSizeExW
	[   7] GetFileVersionInfoSizeW
	[   8] GetFileVersionInfoW
	[   9] VerFindFileA
	[  10] VerFindFileW
	[  11] VerInstallFileA
	[  12] VerInstallFileW
	[  13] VerLanguageNameA
	[  14] VerLanguageNameW
	[  15] VerQueryValueA
	[  16] VerQueryValueW
```